File: app/src/main/java/com/przemolab/oknotifier/modules/OpenKattisService.java
package com.przemolab.oknotifier.modules;

import com.przemolab.oknotifier.models.Contest;
import com.przemolab.oknotifier.models.Contestant;
import com.przemolab.oknotifier.utils.DateUtils;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import timber.log.Timber;

public class OpenKattisService {

    private final String BaseKattisUrl = "https://open.kattis.com";

    public List<Contest> getOngoingContests() {
        try {
            List<Contest> contests = new ArrayList<>();

            String url = String.format("%s/contests", BaseKattisUrl);
            Document contestsPageDocument = Jsoup.connect(url).timeout(0).get();
            Elements elements = contestsPageDocument.select(".main-content table:first-of-type tbody tr");

            for (Element row : elements) {
                Contest contest = getContest(row);
                if (contest != null) {
                    contests.add(contest);
                }
            }

            return contests;
        } catch (Exception ex) {
            Timber.e(ex);
            return null;
        }
    }

    public List<Contestant> getContestStandings(String contestId) {
        try {
            List<Contestant> contestants = new ArrayList<>();

            String url = String.format("%s/contests/%s", BaseKattisUrl, contestId);
            Document contestsPageDocument = Jsoup.connect(url).timeout(0).get();
            Elements elements = contestsPageDocument.select("#standings tbody tr");

            for (Element row : elements) {
                Contestant contestant = getContestant(row, contestId);
                if (contestant != null) {
                    contestants.add(contestant);
                }
            }

            return contestants;
        } catch (Exception ex) {
            Timber.e(ex);
            return null;
        }
    }

    private Contest getContest(Element row) {
        try {
            Elements cells = row.select("td");

            String name = cells.get(0).select("a").text();
            String contestUrl = String.format("%s%s", BaseKattisUrl, cells.get(0).select("a").attr("href"));
            String[] parts = contestUrl.split("/");
            String id = parts[parts.length - 1];

            String startDateText = cells.get(3).text().replace(" CEST", "");

            String dateTimeFormat = "yyyy-MM-dd HH:mm:ss";
            if (startDateText.length() == 8) {
                dateTimeFormat = "HH:mm:ss";
            }
            Date startDate = DateUtils.getDate(startDateText, dateTimeFormat);

            String lengthText = cells.get(2).text();
            Date endDate = DateUtils.addTimeToDate(startDate, lengthText);

            Document standingsPageDocument = Jsoup.connect(contestUrl).timeout(0).get();
            int numberOfContestants = standingsPageDocument.select("#standings tbody tr").size() - 4;
            int numberOfProblems = standingsPageDocument.select("#standings thead th.problemcolheader-standings").size();

            return new Contest(id, name, startDate, endDate, numberOfContestants, numberOfProblems);
        } catch (Exception ex) {
            Timber.e(ex);
            return null;
        }
    }

    private Contestant getContestant(Element row, String contestId) {
        try {
            Elements cells = row.select("td");

            String rank = cells.get(0).select(".rank").text();

            if (!rank.isEmpty()) {

                String name = cells.get(1).select("a").text();
                if (name.isEmpty()) {
                    name = cells.get(1).select("div").text();
                }
                if (name.isEmpty()) {
                    name = cells.get(1).select("em").text();
                }

                int time = Integer.valueOf(cells.get(3).text());

                int solved = 0;
                int submitted = 0;
                int failed = 0;
                int notTried = 0;
                for (Element cell : cells) {
                    if (cell.hasClass("solved")) {
                        solved++;
                    } else if (cell.hasClass("pending")) {
                        submitted++;
                    } else if (cell.hasClass("attempted")) {
                        failed++;
                    } else if (!cell.hasText()) {
                        notTried++;
                    }
                }

                return new Contestant(name, contestId, solved, submitted, failed, notTried, time);
            }

            return null;
        } catch (Exception ex) {
            Timber.e(ex);
            return null;
        }
    }
}

File: app/src/main/java/com/przemolab/oknotifier/modules/OpenKattisService.kt
package com.przemolab.oknotifier.modules

import com.przemolab.oknotifier.data.entries.ContestEntry
import com.przemolab.oknotifier.interfaces.IOpenKattisService
import com.przemolab.oknotifier.models.Contestant
import com.przemolab.oknotifier.utils.DateUtils

import org.jsoup.Jsoup
import org.jsoup.nodes.Element

import java.util.ArrayList

import timber.log.Timber

class OpenKattisService : IOpenKattisService {

    private val _baseKattisUrl = "https://open.kattis.com"

    override val ongoingContests: List<ContestEntry>?
        get() {
            try {
                val contests = ArrayList<ContestEntry>()

                val url = String.format("%s/contests", _baseKattisUrl)
                val contestsPageDocument = Jsoup.connect(url).timeout(0).get()
                val elements = contestsPageDocument.select(".main-content table:first-of-type tbody tr")

                for (row in elements) {
                    val contest = getContest(row)
                    if (contest != null) {
                        contests.add(contest)
                    }
                }

                return contests
            } catch (ex: Exception) {
                Timber.e(ex)
                return null
            }
        }

    override fun getContestStandings(contestId: String): List<Contestant>? {
        try {
            val contestants = ArrayList<Contestant>()

            val url = String.format("%s/contests/%s", _baseKattisUrl, contestId)
            val contestsPageDocument = Jsoup.connect(url).timeout(0).get()
            val elements = contestsPageDocument.select("#standings tbody tr")

            for (row in elements) {
                val contestant = getContestant(row, contestId)
                if (contestant != null) {
                    contestants.add(contestant)
                }
            }

            return contestants
        } catch (ex: Exception) {
            Timber.e(ex)
            return null
        }

    }

    private fun getContest(row: Element): ContestEntry? {
        try {
            val cells = row.select("td")

            val name = cells[0].select("a").text()
            val contestUrl = String.format("%s%s", _baseKattisUrl, cells[0].select("a").attr("href"))
            val parts = contestUrl.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val contestId = parts[parts.size - 1]

            val startDateText = cells[3].text().replace(" CEST", "")

            var dateTimeFormat = "yyyy-MM-dd HH:mm:ss"
            if (startDateText.length == 8) {
                dateTimeFormat = "HH:mm:ss"
            }
            val startDate = DateUtils.getDate(startDateText, dateTimeFormat)

            val lengthText = cells[2].text()
            val endDate = DateUtils.addTimeToDate(startDate, lengthText)

            val standingsPageDocument = Jsoup.connect(contestUrl).timeout(0).get()
            val numberOfContestants = standingsPageDocument.select("#standings tbody tr").size - 4
            val numberOfProblems = standingsPageDocument.select("#standings thead th.problemcolheader-standings").size

            return ContestEntry(contestId = contestId, name = name, startDate = startDate, endDate = endDate,
                    numberOfContestants = numberOfContestants, numberOfProblems = numberOfProblems)
        } catch (ex: Exception) {
            Timber.e(ex)
            return null
        }
    }

    private fun getContestant(row: Element, contestId: String): Contestant? {
        try {
            val cells = row.select("td")

            val rank = cells[0].select(".rank").text()

            if (!rank.isEmpty()) {

                var name = cells[1].select("a").text()
                if (name.isEmpty()) {
                    name = cells[1].select("div").text()
                }
                if (name.isEmpty()) {
                    name = cells[1].select("em").text()
                }

                val time = Integer.valueOf(cells[3].text())

                var solved = 0
                var submitted = 0
                var failed = 0
                var notTried = 0
                for (cell in cells) {
                    if (cell.hasClass("solved")) {
                        solved++
                    } else if (cell.hasClass("pending")) {
                        submitted++
                    } else if (cell.hasClass("attempted")) {
                        failed++
                    } else if (!cell.hasText()) {
                        notTried++
                    }
                }

                return Contestant(name, contestId, solved, submitted, failed, notTried, time)
            }

            return null
        } catch (ex: Exception) {
            Timber.e(ex)
            return null
        }

    }
}